### PR TITLE
feat(pace-83): add paging to external api

### PR DESCRIPTION
**Pull request by trietsch**

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -9,4 +9,4 @@ springCloudKubernetesVersion=3.1.0
 
 dockertag = pace-local
 
-generatedBufDependencyVersion=00000000000000.c11a58c69ce3
+generatedBufDependencyVersion=00000000000000.c46eb729f8cf
```
Comment (bobvandenhoogen) [inline gradle.properties]: Think this needs to be `00000000000000.934197b6cd79`

Comment (bvdeenen) [reply gradle.properties]: have updated already

File: app/src/main/kotlin/com/getstrm/pace/catalogs/DataCatalog.kt [hunk 2]
```diff
@@ -41,7 +53,7 @@ abstract class DataCatalog(
      * A schema is a collection of tables.
      */
     abstract class Schema(val database: Database, val id: String, val name: String) {
-        open suspend fun getTables(): List<Table> = emptyList()
+        open suspend fun listTables(pageParameters: PageParameters = DEFAULT_PAGE_PARAMETERS): PagedCollection<Table> = emptyList<Table>().withPageInfo()
```
Comment (bvdeenen) [inline app/src/main/kotlin/com/getstrm/pace/catalogs/DataCatalog.kt]: should just be abstract

File: app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt [hunk 3]
```diff
@@ -15,22 +22,42 @@ import java.util.*
 class OpenDataDiscoveryCatalog(configuration: CatalogConfiguration) : DataCatalog(configuration) {
     private val searchClient = SearchApi(configuration.serverUrl)
     private val datasetsClient = DataSetApi(configuration.serverUrl)
+    private val dataEntityClient = DataEntityApi(configuration.serverUrl)
```
Comment (bvdeenen) [inline app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt]: unused?

File: app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt [hunk 4]
```diff
@@ -54,9 +93,24 @@ class OpenDataDiscoveryCatalog(configuration: CatalogConfiguration) : DataCatalo
         id: String,
         name: String,
     ) : DataCatalog.Schema(oddDatabase, id, name) {
-        override suspend fun getTables(): List<DataCatalog.Table> {
+        override suspend fun listTables(pageParameters: PageParameters): PagedCollection<DataCatalog.Table> {
```
Comment (bvdeenen) [inline app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt:96]: should be abstract

File: app/src/main/kotlin/com/getstrm/pace/util/Utils.kt [hunk 5]
```diff
@@ -53,3 +59,71 @@ private fun DataPolicy.RuleSet.Filter.listPrincipals() = when (this.filterCase)
     DataPolicy.RuleSet.Filter.FilterCase.GENERIC_FILTER -> this.genericFilter.conditionsList.flatMap { it.principalsList }
     else -> throw IllegalArgumentException("Unsupported filter: ${this.filterCase.name}")
 }
+
+/**
+ * safe page parameters application.
+ * Can't throw an exception
+ */
+fun <T>List<T>.applyPageParameters(p:PageParameters) : List<T> =
```
Comment (bvdeenen) [inline app/src/main/kotlin/com/getstrm/pace/util/Utils.kt:67]: max no longer needed now that skip and page_size are unsigned int 32

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt [hunk 6]
```diff
@@ -123,20 +133,23 @@ class BigQueryClient(
...
-    override suspend fun listGroups(): List<Group> {
+    override suspend fun listGroups(pageParameters: PageParameters): PagedCollection<Group> {
         val query = """
             SELECT
               DISTINCT userGroup
             FROM
               $userGroupsTable
+            ORDER BY userGroup
+            LIMIT ${pageParameters.pageSize}
+            OFFSET ${pageParameters.skip}
```
Comment (bobvandenhoogen) [inline app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt:144]: Have we tested whether or not we can do this with jooq templating? If possible, we should do it here

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt [hunk 7]
```diff
@@ -73,9 +77,16 @@ class SnowflakeClient(
...
         val statement =
-            "select table_schema, table_name, table_type, created as create_date, last_altered as modify_date from information_schema.tables where table_type = 'BASE TABLE';"
+            """SELECT table_schema, table_name, table_type,
+               created as create_date,
+               last_altered as modify_date
+               FROM information_schema.tables where table_type = 'BASE TABLE'
+               ORDER BY table_name
+               LIMIT ${pageParameters.pageSize}
+               OFFSET ${pageParameters.skip} ;
+            """.trimMargin()
```
Comment (bobvandenhoogen) [inline app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt:89]: I think this needs templating

**Review by bobvandenhoogen — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt (hunk 6 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt:144]: no I haven't.

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt (hunk 7 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt:89]: My suggestion is to create a separate ticket for this. It's not needed for the pagination, and this MR is already too big.

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt (hunk 6 above)
Comment (trietsch) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt:144]: Agreed with Bob here. There's no danger of SQL injection whatsoever, as these are `uints`, though jooq templating would be nice.

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt (hunk 7 above)
Comment (trietsch) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt:89]: I do get what you're saying in creating a new ticket for this, but postponing these kind of refactorings will eventually result in not doing them at all. We've all been there.

File: .github/workflows/publish-release-artifacts.yaml [hunk 8]
```diff
@@ -50,7 +50,7 @@ jobs:
           buf generate
       - name: Copy JSON Schema
         run: |
```
Comment (trietsch) [inline .github/workflows/publish-release-artifacts.yaml:52]: Please update this command to reflect the `copy-json-schema-to-resources` makefile task. Even better: use the makefile task itself.

File: app/src/main/kotlin/com/getstrm/pace/util/Utils.kt [hunk 9]
```diff
@@ -53,3 +59,72 @@ private fun DataPolicy.RuleSet.Filter.listPrincipals() = when (this.filterCase)
...
+            .skip(p.skip.toLong())
+            .limit(p.pageSize.toLong())
+            .collect(Collectors.toList())
+
+
+data class PagedCollection<T>(val data: Collection<T>, val pageInfo: PageInfo){
+    fun <V> map(transform: (T) -> V): PagedCollection<V> = data.map(transform).withPageInfo(pageInfo)
+    fun firstOrNull() = data.firstOrNull()
+    fun find(predicate: (T) -> Boolean) = data.find(predicate)
+    fun filter(predicate: (T) -> Boolean) = data.filter(predicate).withPageInfo()
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/util/Utils.kt:79]: I don't think the page info that is set here is correct. Either way, the filtered result should include a new page info that respects the filter.

File: app/src/main/kotlin/com/getstrm/pace/catalogs/Collibra.kt [hunk 10]
```diff
@@ -80,12 +127,27 @@ class CollibraCatalog(config: CatalogConfiguration) : DataCatalog(config) {
...
         stringAttributes.find { it.type.publicId == "DataSourceType" }?.stringValue ?: "unknown"
+
+    private fun GetDataBaseQuery.Asset.getDataSourceType(): String =
+        stringAttributes.find { it.type.publicId == "DataSourceType" }?.stringValue ?: "unknown"
+}
+
+/**
+ * utility function to provide error handling in grpc compatible format.
+ */
+private suspend fun <D : Operation.Data> ApolloCall<D>.executeReturnValidData(): D {
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/catalogs/Collibra.kt]: I'd prefer to name this something like `responseOrError`

File: app/src/main/kotlin/com/getstrm/pace/catalogs/Collibra.kt [hunk 11]
```diff
@@ -1,73 +1,120 @@
...
+                client.query(ListTablesInSchemaQuery(id, skip, pageSize))
+                    .executeReturnValidData().assets.onlyNonNulls().first().tables
+            }.map { Table(catalog, this, it.target.id.toString(), it.target.fullName) }
+            return tables.withUnknownTotals()
+        }
+
+        override suspend fun getTable(tableId: String): DataCatalog.Table {
+            val table = catalog.client.query(GetTableQuery(tableId)).executeReturnValidData().assets.onlyNonNulls().firstOrNull() ?:
+            throw BadRequestException(
+                INVALID_ARGUMENT, BadRequest.getDefaultInstance(),
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/catalogs/Collibra.kt]: In order to provide rich error messages, this `BadRequest` shouldn't be the default instance.

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt (hunk 7 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt:89]: this query is not processed by jooq, it goes directly to SnowFlake...

**Review by trietsch — COMMENTED** [review]
.

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt (hunk 6 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt:144]: this query is not processed by jooq, it goes directly to BigQuery. Does it still make sense to include jooq? In that case with which SQL dialect? BigQuery is only available with the commercial Jooq which we don't (yet) have.

I'm saying we should _not_ do the Jooq templating here.

File: app/src/main/kotlin/com/getstrm/pace/util/Utils.kt (hunk 9 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/util/Utils.kt:79]: it does! Look at the definition of `withPageInfo()` There's even a unit-test for it :-)

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt (hunk 6 above)
Comment (trietsch) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt:144]: Jooq provides typesafety. A plain string containing SQL isn't. I agree with the fact that we don't use commercial Jooq (which we can include in PACE now, as PACE is OS, Jooq Commercial can be used freely IIRC), which prevents us from using the correct dialect.

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt (hunk 7 above)
Comment (trietsch) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt:89]: Same argument as with BigQuery. With Snowflake, we can use the default dialect, which follows ANSI SQL.

Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/snowflake/SnowflakeClient.kt:89]: Created separate ticket https://linear.app/strmprivacy/issue/PACE-88/change-sql-strings-to-jooq-query-builders

File: app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt (hunk 6 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/processing_platforms/bigquery/BigQueryClient.kt:144]: Separate ticket https://linear.app/strmprivacy/issue/PACE-88/change-sql-strings-to-jooq-query-builders

**Pull request merged**